File: triatlon-java/src/main/java/ro/ubbcluj/scs/dfar2166/triatlonjava/repository/db/RefereeDBRepository.java
package ro.ubbcluj.scs.dfar2166.triatlonjava.repository.db;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.ubbcluj.scs.dfar2166.triatlonjava.model.activity.RaceType;
import ro.ubbcluj.scs.dfar2166.triatlonjava.model.person.Referee;
import ro.ubbcluj.scs.dfar2166.triatlonjava.repository.RefereeRepository;
import ro.ubbcluj.scs.dfar2166.triatlonjava.utils.JDBCUtils;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

// RefereeDBRepository

public class RefereeDBRepository implements RefereeRepository {

    // Private Properties

    private final JDBCUtils databaseUtils;

    // Private Class Properties

    private static final Logger logger = LogManager.getLogger();

    // Lifecycle

    public RefereeDBRepository(Properties properties) {
        logger.info("Initialising RefereeDBRepository with properties: {}", properties);
        databaseUtils = new JDBCUtils(properties);
    }

    // Private Methods

    private Referee buildRefereeFromResultSet(ResultSet resultSet) throws SQLException {
        Referee referee = null;
        var id = resultSet.getInt("id");
        var firstName = resultSet.getString("firstName");
        var lastName = resultSet.getString("lastName");
        var raceType = resultSet.getInt("raceType");
        var email = resultSet.getString("email");
        var password = resultSet.getString("password");
        referee = new Referee(firstName, lastName, RaceType.fromValue(raceType), email, password);
        referee.setId(id);
        return referee;
    }

    // Override Methods (RefereeRepository

    @Override
    public Referee findByEmail(String email) {
        logger.traceEntry("Finding task {}", email);
        Connection connection = databaseUtils.getConnection();
        Referee referee = null;
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from referees where email=?")) {
            preparedStatement.setString(1, email);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    referee = buildRefereeFromResultSet(resultSet);
                }
                logger.traceExit("Found 1 instance");
            }
        } catch (SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return referee;
    }

    // Override Methods (Repository)

    @Override
    public Referee findOne(Integer integer) {
        logger.traceEntry("Finding task {}", integer);
        Connection connection = databaseUtils.getConnection();
        Referee referee = null;
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from referees where id=?")) {
            preparedStatement.setInt(1, integer);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    referee = buildRefereeFromResultSet(resultSet);
                }
                logger.traceExit("Found 1 instance");
            }
        } catch (SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return referee;
    }

    @Override
    public Iterable<Referee> findAll() {
        logger.traceEntry("Finding task");
        Connection connection = databaseUtils.getConnection();
        List<Referee> referees = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement("select * from referees")) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Referee referee = buildRefereeFromResultSet(resultSet);
                    referees.add(referee);
                }
                logger.traceExit("Found {} instances", referees.size());
            }
        } catch (SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return referees;
    }

    @Override
    public Referee save(Referee entity) {
        logger.traceEntry("Saving task {}", entity);
        Connection connection = databaseUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("insert into referees (firstName, lastName, raceType, email, password) values (?, ?, ?, ?, ?)")) {
            preparedStatement.setString(1, entity.getFirstName());
            preparedStatement.setString(2, entity.getLastName());
            preparedStatement.setInt(3, entity.getRaceType().getValue());
            preparedStatement.setString(4, entity.getEmail());
            preparedStatement.setString(5, entity.getPassword());
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Saved {} instances", result);
        } catch (SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return entity;
    }

    @Override
    public Integer delete(Integer integer) {
        logger.traceEntry("Deleting task {}", integer);
        Connection connection = databaseUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("delete from referees where id=?")) {
            preparedStatement.setInt(1, integer);
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Deleted {} instances", result);
        } catch (SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return integer;
    }

    @Override
    public Referee update(Referee entity) {
        logger.traceEntry("Updating task {}", entity);
        Connection connection = databaseUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("update referees set firstName=?, lastName=?, raceType=?, email=?, password=? where id=?")) {
            preparedStatement.setString(1, entity.getFirstName());
            preparedStatement.setString(2, entity.getLastName());
            preparedStatement.setInt(3, entity.getRaceType().getValue());
            preparedStatement.setString(4, entity.getEmail());
            preparedStatement.setString(5, entity.getPassword());
            preparedStatement.setInt(6, entity.getId());
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Updated {} instances", result);
        } catch (SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return entity;
    }

}

File: triatlon-cs/Repository/DB/RefereeDbRepository.cs
using System.Collections.Generic;
using System.Data;
using log4net;
using Model.Activity;
using Model.Person;
using Repository.Utils;

namespace Repository.DB
{
    public class RefereeDbRepository : IRefereeRepository
    {
    
        // Private Properties

        private static readonly ILog Logger = LogManager.GetLogger("RefereeDbRepository");
        private readonly IDictionary<string, string> _properties;
        
        // Lifecycle

        public RefereeDbRepository(IDictionary<string, string> properties)
        {
            Logger.Info("Creating RefereeDbRepository");
            _properties = properties;
        }
        
        // Repository Methods (IRefereeRepository)

        public Referee FindByEmail(string email)
        {
            Logger.InfoFormat("Entering FindEmail with value {0}", email);
            IDbConnection connection = DbUtils.GetConnection(_properties);
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from referees where email=@email";
                var parameterId = command.CreateParameter();
                parameterId.ParameterName = "@email";
                parameterId.Value = email;
                command.Parameters.Add(parameterId);
                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        int _id = dataReader.GetInt32(0);
                        string firstName = dataReader.GetString(1);
                        string lastName = dataReader.GetString(2);
                        RaceType raceType = (RaceType) dataReader.GetInt32(3);
                        string password = dataReader.GetString(5);
                        var referee = new Referee(firstName, lastName, raceType, email, password);
                        referee.Id = _id;
                        Logger.InfoFormat("Exiting FindOne with value {0}", referee);
                        return referee;
                    }
                }
            }
            Logger.InfoFormat("Exiting FindOne with value {0}", null);
            return null;
        }
        
        // Repository Methods
        
        public Referee FindOne(int id)
        {
            Logger.InfoFormat("Entering FindOne with value {0}", id);
            IDbConnection connection = DbUtils.GetConnection(_properties);
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from referees where id=@id";
                var parameterId = command.CreateParameter();
                parameterId.ParameterName = "@id";
                parameterId.Value = id;
                command.Parameters.Add(parameterId);
                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        int _id = dataReader.GetInt32(0);
                        string firstName = dataReader.GetString(1);
                        string lastName = dataReader.GetString(2);
                        RaceType raceType = (RaceType) dataReader.GetInt32(3);
                        string email = dataReader.GetString(4);
                        string password = dataReader.GetString(5);
                        var referee = new Referee(firstName, lastName, raceType, email, password);
                        referee.Id = _id;
                        Logger.InfoFormat("Exiting FindOne with value {0}", referee);
                        return referee;
                    }
                }
            }
            Logger.InfoFormat("Exiting FindOne with value {0}", null);
            return null;
        }

        public IEnumerable<Referee> FindAll()
        {
            Logger.Info("Entering FindAll");
            IDbConnection connection = DbUtils.GetConnection(_properties);
            IList<Referee> referees = new List<Referee>();
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from referees";
                using (var dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        int _id = dataReader.GetInt32(0);
                        string firstName = dataReader.GetString(1);
                        string lastName = dataReader.GetString(2);
                        RaceType raceType = (RaceType) dataReader.GetInt32(3);
                        string email = dataReader.GetString(4);
                        string password = dataReader.GetString(5);
                        var referee = new Referee(firstName, lastName, raceType, email, password);
                        referee.Id = _id;
                        referees.Add(referee);
                    }
                }
            }
            Logger.Info("Exiting FindAll");
            return referees;
        }

        public Referee Save(Referee entity)
        {
            Logger.InfoFormat("Entering Save with value {0}", entity);
            IDbConnection connection = DbUtils.GetConnection(_properties);
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "insert into referees (firstName, lastName, raceType, email, password) values (@firstName, @lastName, @raceType, @email, @password)";
                var firstNameParameter = command.CreateParameter();
                firstNameParameter.ParameterName = "@firstName";
                firstNameParameter.Value = entity.FirstName;
                command.Parameters.Add(firstNameParameter);
                var lastNameParameter = command.CreateParameter();
                lastNameParameter.ParameterName = "@lastName";
                lastNameParameter.Value = entity.LastName;
                command.Parameters.Add(lastNameParameter);
                var raceTypeParameter = command.CreateParameter();
                raceTypeParameter.ParameterName = "@raceType";
                raceTypeParameter.Value = (int) entity.RaceType;
                command.Parameters.Add(raceTypeParameter);
                var emailParameter = command.CreateParameter();
                emailParameter.ParameterName = "@email";
                emailParameter.Value = entity.Email;
                command.Parameters.Add(emailParameter);
                var passwordParameter = command.CreateParameter();
                passwordParameter.ParameterName = "@password";
                passwordParameter.Value = entity.Password;
                command.Parameters.Add(passwordParameter);
                var result = command.ExecuteNonQuery();
                Logger.InfoFormat("Exiting Save with value {0}", result);
            }
            return entity;
        }

        public int Delete(int id)
        {
            Logger.InfoFormat("Entering Delete with value {0}", id);
            IDbConnection connection = DbUtils.GetConnection(_properties);
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "delete from referees where id=@id";
                var idParameter = command.CreateParameter();
                idParameter.ParameterName = "@id";
                idParameter.Value = id;
                command.Parameters.Add(idParameter);
                var result = command.ExecuteNonQuery();
                Logger.InfoFormat("Exiting Delete with value {0}", result);
            }
            return id;
        }

        public Referee Update(Referee entity)
        {
            Logger.InfoFormat("Entering Update with value {0}", entity);
            IDbConnection connection = DbUtils.GetConnection(_properties);
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "update referees set firstName=@firstName, lastName=@lastName, raceType=@raceType, email=@email, password=@password where id=@id";
                var idParameter = command.CreateParameter();
                idParameter.ParameterName = "@id";
                idParameter.Value = entity.Id;
                command.Parameters.Add(idParameter);
                var firstNameParameter = command.CreateParameter();
                firstNameParameter.ParameterName = "@firstName";
                firstNameParameter.Value = entity.FirstName;
                command.Parameters.Add(firstNameParameter);
                var lastNameParameter = command.CreateParameter();
                lastNameParameter.ParameterName = "@lastName";
                lastNameParameter.Value = entity.LastName;
                command.Parameters.Add(lastNameParameter);
                var raceTypeParameter = command.CreateParameter();
                raceTypeParameter.ParameterName = "@raceType";
                raceTypeParameter.Value = (int) entity.RaceType;
                command.Parameters.Add(raceTypeParameter);
                var emailParameter = command.CreateParameter();
                emailParameter.ParameterName = "@email";
                emailParameter.Value = entity.Email;
                command.Parameters.Add(emailParameter);
                var passwordParameter = command.CreateParameter();
                passwordParameter.ParameterName = "@password";
                passwordParameter.Value = entity.Password;
                command.Parameters.Add(passwordParameter);
                var result = command.ExecuteNonQuery();
                Logger.InfoFormat("Exiting Update with value {0}", result);
            }
            return entity;
        }
    
    }
}